Symbiota/Symbiota
# Add Audio File Support

# Issues #1389 #1554 #1586 #1587 #1588 #1599 #1600 #1601 #1692 #1693 #1694 #1696 #1697 #1698 #1699 #1673 #1711

This is more of a draft pr currently to get the review process started because looking at the file changed count it is gonna take a while to get it though.

# Summary
This is a total overhaul of the image systems to be generalized to any media with a current focus on audio but with video support in mind. This is a very large pr because it is a very large overhaul of deeply embedded systems

## Design
The general idea was to make a static class that was easy to use for the manager classes and other misc classes. Having it extend manager would be a hassle if their was a manager being injected into another manager or dealing with more polymorphic managers when executing a variety of post actions. Since it was static some utility classes needed to be made for cleaning input and output as well as database connections.

## Backwards Compatibility Issues
I have gone back and retroactively fixed most of the backwards compatibility issues that were introduced prior to our discussion on the topic. There might be some I missed though.

## Classes
- `Media` : Handles crud operations on media table and its related tables
- `MediaException` : For media specific errors commonly used to bubble up status messages for the ui
- `StorageStrategy`: Class aimed at getting our file storage logic under a single api that could be generalized for remote storage options.
- `LocalStorage`: Extends StorageStrategy as is used for storing, deleting, renaming, and remapping files on local storage
- `Sanitize`: Sanitizing utility that moves all the Manager sanitation logic into a static class with a simple api, only `out` for output and `in` for input sanitation. This was talked about a while ago and I figured I would take a crack at implementing it since it was relevant to work I was already doing.
- `Database`: Allows for composable database connection access for static classes without need for inheritance.

A lot of these classes are in the media.php file and I think that make sense for now, but I would be happy to split them into separate files if their is a strong opposition to how things are organized.

## Language Files
There is a lot of language file changes that Katie and others have worked on reviewing I would ignore them when doing the code review. 

## Deprecating
### OccurrenceEditorImages
Shouldn't be used for media upload and edits anymore
Still Used:
- `webservices/imageuploader.php` - Not used anywhere else could this be removed?
- `imageoccursubmit.php` - No plans on updating this to use `Media` class but could if desired.

### ImageDetailManager
Shouldn't be used moving forward
Still used:
- `imgdetails.php` - only for `getDomain` function could easily get removed if the function was made available via utilities.
- `ActionManager` - Not really sure what this is did not touch.

### Not Deprecated but could be in the future
- `ImageShared.php` - Touched to many other parts of the code base so didn't implement `Media` class in place of it to save time and scope creep.

### How to Test
Media could be image or audio both should be tested to make sure system work with both
- Occurrence Editor
  - Upload from media
  - Upload from remote
  - Upload Larger file than max upload size (should give you message)
  - Update Metadata
  - Disassociate media
  - Remap media
  - Delete media
- Taxon Profile
  - Check that taxa profile page has audio displayed in media gallery as a speaker icon. Should not be able to play audio from taxon profile as of now.
  - Edit Media Metadata from taxon profile admin page (Should take you media detail page)
- Media Detail
  - Update
  - Unlink from taxon
  - Delete
- Media Search
  - Search for only audio
  - Search for only images
- Occurrence Profile
  - Uploaded Audio should be displayed on occurrence page and be playable
- Upload
  - Audio can be uploaded via collprofile -> Import/Update Specimen Records -> Extended Data Import 
- Download 
  - any download with media options selected. Make sure exported links are correct for hosted media and external media

# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [x] Hotfixes should be branched off of the `master` branch and **squash and merged** back into the `master` branch.
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [x] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [x] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [x] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
- [x] Comment which GitHub issue(s), if any does this PR address
- [x] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).

# Post-Approval

- [x] It is the code author's responsibility to merge their own pull request after it has been approved
- [x] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [x] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [x] If this PR represents a hotfix into the `master` branch, a subsequent PR from `master` into `Development` should be made **merge** option (i.e., no squash).
- [x] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [x] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!
